router/sleeve.go

```diff
@@ -0,0 +1,1117 @@
+
+	if fwd.heartbeatInterval != 0 {
+		// already confirmed
+		return nil
+	}
+
+	// heartbeatInterval flags that we want to send heartbeats,
+	// even if we don't do sendHeartbeat() yet due to lacking the
+	// remote address.
+	fwd.heartbeatInterval = time.Duration(FastHeartbeat)
+	if fwd.remoteAddr != nil {
+		if err := fwd.sendHeartbeat(); err != nil {
+			return err
+		}
+	}
+
+	fwd.heartbeatTimeout = time.NewTimer(HeartbeatTimeout)
+	return nil
+}
+
+func (fwd *sleeveForwarder) sendHeartbeat() error {
+	log.Debug(fwd.logPrefix(), "sendHeartbeat")
+
+	// Prime the timer for the next heartbeat.  We don't use a
+	// ticker because the interval is not constant.
+	fwd.heartbeatTimer = setTimer(fwd.heartbeatTimer, fwd.heartbeatInterval)
+
+	buf := make([]byte, EthernetOverhead+8)
+	binary.BigEndian.PutUint64(buf[EthernetOverhead:], fwd.connUID)
+	return fwd.sendSpecial(fwd.crypto.EncDF, fwd.senderDF, buf)
+}
+
+func (fwd *sleeveForwarder) handleHeartbeat(special specialFrame) error {
+	uid := binary.BigEndian.Uint64(special.frame[EthernetOverhead:])
+	if uid != fwd.connUID {
+		return nil
+	}
+
+	log.Debug(fwd.logPrefix(), "handleHeartbeat")
+
+	if fwd.remoteAddr == nil {
+		fwd.setRemoteAddr(special.sender)
+		if fwd.heartbeatInterval != time.Duration(0) {
+			if err := fwd.sendHeartbeat(); err != nil {
+				return err
+			}
+		}
+	} else if !udpAddrsEqual(fwd.remoteAddr, special.sender) {
+		log.Println(fwd.logPrefix(),
+			"Peer UDP address changed to", special.sender)
+		fwd.setRemoteAddr(special.sender)
+	}
+
+	if !fwd.ackedHeartbeat {
+		fwd.ackedHeartbeat = true
+		if err := fwd.sendControlMsg([]byte{HeartbeatAck}); err != nil {
+			return err
+		}
+	}
+
+	fwd.heartbeatTimeout.Reset(HeartbeatTimeout)
+	return nil
+}
+
+func (fwd *sleeveForwarder) setRemoteAddr(addr *net.UDPAddr) {
+	// Although we don't need to lock when reading remoteAddr,
+	// because this thread is the only one that modifies
+	// remoteAddr, we do need to lock when writing it, because
+	// memory models.
+	fwd.lock.Lock()
+	fwd.remoteAddr = addr
+	fwd.lock.Unlock()
+}
+
+func (fwd *sleeveForwarder) handleHeartbeatAck() error {
+	// The connection is nowregarded as established
+	fwd.notifyEstablished()
+
+	if fwd.heartbeatInterval != SlowHeartbeat {
+		fwd.heartbeatInterval = SlowHeartbeat
+		if fwd.heartbeatTimer != nil {
+			fwd.heartbeatTimer.Reset(fwd.heartbeatInterval)
+		}
+	}
+
+	fwd.fragTestTicker = time.NewTicker(FragTestInterval)
+	if err := fwd.sendFragTest(); err != nil {
+		return err
+	}
+
+	// Send a large frame down the DF channel in order to prompt
+	// PMTU discovery to start.
+	return fwd.sendSpecial(fwd.crypto.EncDF, fwd.senderDF,
+		make([]byte, PMTUDiscoverySize))
+}
+
+func (fwd *sleeveForwarder) notifyEstablished() {
+	fwd.lock.RLock()
+	defer fwd.lock.RUnlock()
+	if fwd.listener != nil {
+		fwd.listener.Established()
+	}
+}
+
+func (fwd *sleeveForwarder) sendFragTest() error {
+	log.Debug(fwd.logPrefix(), "sendFragTest")
+	fwd.stackFrag = false
```

Comment:
> From linked doc "Reads and writes of values larger than a single machine word behave as multiple machine-word-sized operations in an unspecified order." implies you are correct.

It's a shame it's not explicit about bools though.  They can be a special case, because for example, if you have `struct { bool a, bool b }` those two fields could in principle be packed within a single byte (as with C bit fields).  But then assignments like `s.a = true` can end up compiling to the equivalent of `*s |= 1`, and so assignments are not atomic.  But then pointers to bit-packed fields need special handling, so I think it's unlikely that any golang implementation would do that even aside from the atomicity concerns.